cargo-miri/bin.rs

```diff
@@ -585,6 +614,34 @@ fn phase_cargo_rustc(args: env::Args) {
         // (Need to do this here as cargo moves that "binary" to a different place before running it.)
         info.store(&out_filename("", ".exe"));
 
+        // Rustdoc expects us to exit with an error code if the test is marked as `compile_fail`,
+        // so we need to run Miri with `MIRI_BE_RUSTC` for a check-only build.
+        if std::env::var_os("MIRI_CALLED_FROM_RUSTDOC").is_some() {
+            let mut cmd = miri();
+
+            // use our own sysroot
+            if !has_arg_flag("--sysroot") {
+                let sysroot = env::var_os("MIRI_SYSROOT")
+                    .expect("the wrapper should have set MIRI_SYSROOT");
+                cmd.arg("--sysroot").arg(sysroot);
+            }
+            
+            // don't go into "code generation" (i.e. validation)
+            if info.args.iter().position(|arg| arg.starts_with("--emit=")).is_none() {
```

Comment:
Can't we use one of the arg-parsing helper functions for this?

Reply:
Also, will this ever be `some`? Might be worth asserting this, since what we certainly do not want is for this to produce a binary that overwrites the JSON file...

Reply:
As of right now, rustdoc only ever passes `--emit=metadata`, but I did add an assert there.

The arg-parsing helpers aren't applicable here, because they look for exact matches, rather than a prefix.

Reply:
Checking the `--emit` flag with the helpers should work though -- the helpers *do* support `=`.

Reply:
Oh the problem is that you are working on `info.args`, not the args of the current process,,, makes sense. But a brief comment along those lines would be a good idea I think.

Reply:
> working on `info.args`, not the args of the current process,,,

Would [`ArgFlagValueIter`](https://github.com/rust-lang/miri/blob/853254f467e26ee1d5d46b7079c96c4f53f072a1/cargo-miri/bin.rs#L102-L103) from #1675 help? That accepts any iterator (of type `TakeWhile<env::Args, fn(&String) -> bool>` currently

<details>
<summary>but can be changed to a generic type)</summary>

```diff
--- a/cargo-miri/bin.rs
+++ b/cargo-miri/bin.rs
@@ -100,12 +100,12 @@ fn has_arg_flag(name: &str) -> bool {
 }
 
 /// Yields all values of command line flag `name`.
-struct ArgFlagValueIter<'a> {
-    args: TakeWhile<env::Args, fn(&String) -> bool>,
+struct ArgFlagValueIter<'a, I> {
+    args: I,
     name: &'a str,
 }
 
-impl<'a> ArgFlagValueIter<'a> {
+impl<'a> ArgFlagValueIter<'a, TakeWhile<env::Args, fn(&String) -> bool>> {
     fn new(name: &'a str) -> Self {
         Self {
             // Stop searching at `--`.
@@ -115,7 +115,7 @@ impl<'a> ArgFlagValueIter<'a> {
     }
 }
 
-impl Iterator for ArgFlagValueIter<'_> {
+impl<I: Iterator<Item = String>> Iterator for ArgFlagValueIter<'_, I> {
     type Item = String;
 
     fn next(&mut self) -> Option<Self::Item> {
```
</details>

Reply:
Yeah, we'd have to generalize, but then this should work. We should also generalize `ArgFlagValueIter::new` then.